File: 397_longest-increasing-continuous-subsequence/longest-increasing-continuous-subsequence.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   taoleetju
@Problem:  http://www.lintcode.com/problem/longest-increasing-continuous-subsequence
@Language: Python
@Datetime: 15-10-06 17:06
'''

class Solution:
    # @param {int[]} A an array of Integer
    # @return {int}  an integer
    def longestIncreasingContinuousSubsequence(self, A):
        # Write your code here
        l = len(A)
        if l < 2:
            return l
        up = True if A[1] > A[0] else False
        current = 2
        longest = 2
        for i in range(2, l):
            if up:
                if A[i] > A[i-1]:
                    current += 1
                else:
                    current = 2
                    up = False
            else:
                if A[i] < A[i-1]:
                    current += 1
                else:
                    current = 2
                    up = True
            longest = current if (current > longest) else longest
        return longest

File: 397_longest-increasing-continuous-subsequence/longest-increasing-continuous-subsequence.cpp
/*
@Copyright:LintCode
@Author:   taoleetju
@Problem:  http://www.lintcode.com/problem/longest-increasing-continuous-subsequence
@Language: C++
@Datetime: 16-07-20 02:27
*/

class Solution {
public:
    /**
     * @param A an array of Integer
     * @return  an integer
     */
    int longestIncreasingContinuousSubsequence(vector<int>& A) {
        int a = lics(A);
        reverse(A.begin(), A.end());
        int b = lics(A);
        return max(a, b);
    }
    
private:
    int lics( vector<int> A){
        if(A.empty())
            return 0;
        vector<int> l(A.size(), 1);
        int max = 1;
        for(int i=1; i<A.size(); i++){
            l[i] = A[i]>A[i-1] ? l[i-1]+1: l[i];
            if(l[i]>max)
                max = l[i];
        }
        return max;
    }
};